### ISSUE TITLE
[R] Is the mxnet always minimize the custom loss function?

### ISSUE THREAD

**Issue by yiyusheng** [apache/incubator-mxnet#4166]
Package used (Python/R/Scala/Julia): R
Before I write my own custom loss function with mx.metric.custom, I checked the `https://github.com/dmlc/mxnet/blob/master/R-package/R/metric.R`.
I find two functions: mx.metric.accuracy is used as a metric for classification and mx.metric.rmse is used as a metric for regression. 
So far as I know, the model should minimize the loss function like MSE. But minimize accuracy indicates a bad model. When I test the mx.metric.accuracy, the increasing accuracy of each epoch indicates the mxnet know how to treat the loss function.
This confuses me when I write my own loss function. I do not  know when the mxnet will maximum the loss function and when the mxnet will minimize the loss function. Could you explain this stuff? Thank you.

**Comment by back2yes** [comment]
I guess adding a minus '-' in front of the loss function would turn the minimization into maxmization.

**Comment by yiyusheng** [comment]
@back2yes , yes.
But I am not sure that the mxnet will always minimize the loss function. Because I find one `mx.metric.accuracy` which should be maximized.

**Comment by marugari** [comment]
`mx.metric` are not loss functions to optimize, but for evaluation.

In your case, MXNet minimized log_loss (If you choose SoftmaxOutput).
It follows that the accuracy score increases.

**Comment by back2yes** [comment]
@yiyusheng I understand you. @marugari was right.
The 'metric' is just a score, it is evaluated after the weights and biases are updated.
Whereas the loss function is calculated before the update, the backward process is based on the loss function value.

**Comment by yiyusheng** [comment]
@marugari, thank you for correcting my fault. I'd like to use my own loss function. Do you know how to do that?

**Comment by yiyusheng** [comment]
@back2yes, yes, I recognize the mistake, thanks. Is there any idea to implement my own loss function?

**Comment by marugari** [comment]
1. add a new Output layer (C++)
https://github.com/dmlc/mxnet/blob/e7514fe1b3265aaf15870b124bb6ed0edd82fa76/src/operator/softmax_output-inl.h

2. use `mx.sym.MakeLoss` (Python/R? I don't know details)
https://github.com/dmlc/mxnet/issues/3368

**Comment by thirdwing** [comment]
In R, you can use `mx.symbol.MakeLoss`

**Comment by thirdwing** [comment]
@yiyusheng You should be able to find some simple examples in https://github.com/dmlc/mxnet/blob/master/docs/tutorials/r/CustomLossFunction.md

**Comment by yiyusheng** [comment]
@thirdwing thank you for your advice. I have sent an e-mail to you to detail my question. Could you have a look on it? Thank you very much.

**Comment by thirdwing** [comment]
The document has been updated. Please see https://github.com/apache/incubator-mxnet/blob/master/docs/tutorials/r/CustomLossFunction.md
